Beta hotspot checklist

danielcolinjames opened the pull request:
screen recording of it in action:
![2021-06-02 20 52 10](https://user-images.githubusercontent.com/10648471/120584872-98000f80-c3e5-11eb-889a-4945177064c4.gif)

had to pretty much completely refactor my old somewhat-spaghetti checklist code. there's still some messy bits in here, will do a bit more tidying up tomorrow

allenan commented on components/InfoBox/HotspotDetails/StatisticsPane.js:
```diff
@@ -19,6 +21,9 @@ const StatisticsPane = ({ hotspot }) => {
     'week',
   )
+
+  const { height, isLoading: heightLoading } = useBlockHeight()
```
since height is only used within the checklist widget, does it make more sense to use the hook within that widget instead of adding more requirements to the calling component?

danielcolinjames replied:
oh yup, definitely, good catch

allenan commented on components/Widgets/ChecklistWidget.js:
```diff
@@ -0,0 +1,229 @@
+const ChecklistWidget = ({ hotspot, witnesses, height }) => {
+  const [activity, setActivity] = useState({})
+  const [loading, setActivityLoading] = useState(true)
+  const [showChecklist, setShowChecklist] = useState(false)
```
another usecase for `useToggle`?

allenan commented on components/Widgets/ChecklistWidget.js:
```diff
@@ -0,0 +1,229 @@
+    const SKELETON_SLICES = 7
+    const sliceWidth = `${(1 / SKELETON_SLICES) * 100}%`
+
+    return (
```
maybe this could be a separate component so that this component reads clearer?

allenan commented on components/Widgets/ChecklistWidget.js:
```diff
@@ -0,0 +1,229 @@
+      <div className="flex items-center justify-center h-10 md:h-5">
+        {sortChecklistItems(possibleChecklistItems).map(
+          (checklistItem, index, { length }) => {
+            const isNextMilestone =
```
given that there are consts being defined in this map, that's a good sign that this should be broken into its own component as well

allenan commented on data/checklist.js:
```diff
@@ -0,0 +1,170 @@
+
+export const getActivityForChecklist = async (address) => {
+  // Get most recent challenger transaction
+  const challengerTxnList = await client.hotspot(address).activity.list({
```
can you parallelize any of the fetching logic in this function with a `Promise.all`?

allenan commented on data/checklist.js:
```diff
@@ -0,0 +1,170 @@
+        isNaN(hotspot.status.height) || isNaN(height)
+          ? `Hotspot is not yet synced.`
+          : height - hotspot.status.height < 500
+          ? `Hotspot is fully synced.`
```
nested ternaries are very hard to read

danielcolinjames replied:
ah yeah I didn't edit any of the existing checklist logic but it can definitely be cleaned up in a number of places

allenan replied:
yeah figured that was just coming over as-is. np. we should also think about extracting text-heavy stuff like this into our i18n file to get ahead of that requirement

danielcolinjames replied:
totally

danielcolinjames commented:
thanks for all the good feedback @allenan, addressed all your comments 👍   and added a show/hide details button:
![2021-06-03 16 08 13](https://user-images.githubusercontent.com/10648471/120723144-01cff600-c486-11eb-81a1-9f82528cf92e.gif)

allenan commented on data/checklist.js:
```diff
@@ -0,0 +1,176 @@
+    }),
+  ])
+  const [challengerTxn, challengeeTxn, rewardTxns] = await Promise.all([
+    await challengerTxnList.take(1),
```
i don't think you need awaits inside the array, it should be implied

danielcolinjames replied:
oh woops yeah you're right

allenan reviewed (approved):
🔥
